src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+
+    // Check if dentry already exists
+    // ==============================
+    InodeStore existing_dentry;
+    r = read_dentry(parent_ino, fragment, dname, &existing_dentry);
+    bool write_dentry = false;
+    if (r == -ENOENT || r == -EINVAL) {
+      // Missing or corrupt dentry
+      write_dentry = true;
+    } else if (r < 0) {
+      derr << "Unexpected error reading dentry 0x" << std::hex
+        << parent_ino << std::dec << "/"
+        << dname << ": " << cpp_strerror(r) << dendl;
+      break;
+    } else {
+      // Dentry already present, does it link to me?
+      if (existing_dentry.inode.ino == ino) {
+        dout(20) << "Dentry 0x" << std::hex
+          << parent_ino << std::dec << "/"
+          << dname << " already exists and points to me" << dendl;
+        // TODO: a tool or frag to overwrite the 'size' in the
+        // existing metadata with the one that we have just recovered
+        // from scan_extents phase
+      } else {
+        // FIXME: at this point we should set a flag to recover
+        // this inode in a /_recovery/<inodeno>.data file as we
+        // can't recover it into its desired filesystem position.
+        derr << "Dentry 0x" << std::hex
+          << parent_ino << std::dec << "/"
+          << dname << " already exists but points to 0x"
+          << std::hex << existing_dentry.inode.ino << std::dec << dendl;
+        // Fall back to lost+found!
+        break;
+      }
+    }
+
+    // Inject linkage
+    // ==============
+    if (write_dentry) {
+      InodeStore dentry;
+      if (i == backtrace.ancestors.begin()) {
+        // This is the linkage for a file
+        dentry.inode.mode = 0500 | S_IFREG;
+        dout(10) << "Linking inode 0x" << std::hex << ino
+          << " at 0x" << parent_ino << "/" << dname << std::dec
+          << " with size=" << file_size << " bytes" << dendl;
+
+        // The file size and mtime we learned by scanning globally
+        dentry.inode.size = file_size;
+        dentry.inode.max_size_ever = file_size;
+        dentry.inode.mtime.tv.tv_sec = file_mtime;
+        dentry.inode.atime.tv.tv_sec = file_mtime;
+        dentry.inode.ctime.tv.tv_sec = file_mtime;
+
+        dentry.inode.layout = g_default_file_layout;
+        dentry.inode.layout.fl_object_size = chunk_size;
+        dentry.inode.layout.fl_stripe_unit = chunk_size;
+        dentry.inode.layout.fl_pg_pool = data_pool_id;
+
+        dentry.inode.truncate_seq = 1;
+        dentry.inode.truncate_size = -1ull;
+
+        dentry.inode.inline_data.version = CEPH_INLINE_NONE;
+      } else {
+        // This is the linkage for a directory
+        dentry.inode.mode = 0755 | S_IFDIR;
+
+        // Set nfiles to something non-zero, to fool any other code
+        // that tries to ignore 'empty' directories.  This won't be
+        // accurate, but it should avoid functional issues.
+        dentry.inode.dirstat.nfiles = 1;
+        dentry.inode.size = 1;
```

Comment:
I think we need a way of marking these as recovering directories, so that we can rebuild their size counts. Unless we already have something like that?